node/src/actors/chain_manager/mod.rs

```diff
@@ -2345,7 +2340,7 @@ impl ChainManager {
         let protocol_version = ProtocolVersion::from_epoch(block.block_header.beacon.checkpoint);
         let replication_factor = self
             .consensus_constants_wit2
-            .get_replication_factor(block.block_header.beacon.checkpoint);
+            .get_replication_factor(block.block_header.beacon.checkpoint, chain_beacon.checkpoint);
```

Comment:
Need to subtract 1
```rust
// The beacon was updated with the current epoch prior to calling this function to build the block
// currently under validation, so subtract one from the beacon to get the prior epoch
let replication_factor = self
    .consensus_constants_wit2
    .get_replication_factor(block.block_header.beacon.checkpoint, chain_beacon.checkpoint - 1);
```

Reply:
Hmm, shouldn't we rather pass `chain_beacon` instead of `beacon` when calling `future_process_validation` from https://github.com/witnet/witnet-rust/blob/master/node/src/actors/chain_manager/mining.rs#L256?

Reply:
I just realized above proposal does not make any sense because this implies that when validating we are always assuming blocks are consecutive which is the whole thing this patch attempts to solve.

Instead, we need something like this:
```
diff --git a/node/src/actors/chain_manager/mining.rs b/node/src/actors/chain_manager/mining.rs
index ecc0e1c9..e4ce6d7f 100644
--- a/node/src/actors/chain_manager/mining.rs
+++ b/node/src/actors/chain_manager/mining.rs
@@ -110,6 +110,7 @@ impl ChainManager {
         let checkpoint_zero_timestamp = chain_info.consensus_constants.checkpoint_zero_timestamp;
         let halving_period = chain_info.consensus_constants.halving_period;

+        let previous_block_epoch = chain_info.highest_block_checkpoint;
         let mut beacon = chain_info.highest_block_checkpoint;
         let mut vrf_input = chain_info.highest_vrf_output;
         let protocol_version = get_protocol_version(self.current_epoch);
@@ -252,6 +253,7 @@ impl ChainManager {
             .and_then(move |block, act, _ctx| {
                 act.future_process_validations(
                     block.clone(),
+                    previous_block_epoch,
                     current_epoch,
                     vrf_input,
                     beacon,
diff --git a/node/src/actors/chain_manager/mod.rs b/node/src/actors/chain_manager/mod.rs
index 7251c626..b4a7a4eb 100644
--- a/node/src/actors/chain_manager/mod.rs
+++ b/node/src/actors/chain_manager/mod.rs
@@ -1239,6 +1239,7 @@ impl ChainManager {
                     // IMPORTANT: Always perform age resets after adding rewards

                     // Reset mining power to block proposer, and higher ranked eligible candidates:
+                    log::info!("Resetting mining age for {} to {}", miner_pkh, block_epoch);
                     let _ = stakes.reset_mining_age(miner_pkh, block_epoch);

                     // Reset witnessing power
@@ -2332,6 +2333,7 @@ impl ChainManager {
     pub fn future_process_validations(
         &mut self,
         block: Block,
+        previous_block_epoch: Epoch,
         current_epoch: Epoch,
         vrf_input: CheckpointVRF,
         chain_beacon: CheckpointBeacon,
@@ -2349,7 +2351,7 @@ impl ChainManager {
         // currently under validation, so subtract one from the beacon to get the prior epoch
         let replication_factor = self
             .consensus_constants_wit2
-            .get_replication_factor(block.block_header.beacon.checkpoint, chain_beacon.checkpoint - 1);
+            .get_replication_factor(block.block_header.beacon.checkpoint, previous_block_epoch);
         let res = validate_block(
             &block,
             current_epoch,
```

Reply:
Yes that makes much more sense.